staging/src/k8s.io/client-go/tools/remotecommand/remotecommand_test.go

```diff
@@ -0,0 +1,238 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remotecommand
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/httpstream/spdy"
+	remotecommandconsts "k8s.io/apimachinery/pkg/util/remotecommand"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+	"time"
+)
+
+type AttachFunc func(in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan TerminalSize) error
+type context struct {
+	conn         io.Closer
+	stdinStream  io.ReadCloser
+	stdoutStream io.WriteCloser
+	stderrStream io.WriteCloser
+	writeStatus  func(status *apierrors.StatusError) error
+}
+
+type streamAndReply struct {
+	httpstream.Stream
+	replySent <-chan struct{}
+}
+
+type fakeMassiveDataPty struct{}
+
+func (s *fakeMassiveDataPty) Read(p []byte) (int, error) {
+	time.Sleep(time.Duration(1) * time.Second)
+	return copy(p, []byte{}), errors.New("client crashed after 1 second")
+}
+
+func (s *fakeMassiveDataPty) Write(p []byte) (int, error) {
+	time.Sleep(time.Duration(1) * time.Second)
+	return len(p), errors.New("return err")
+}
+
+func fakeMassiveDataAttacher(in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan TerminalSize) error {
+
+	stdinDone := make(chan struct{})
+	stdoutDone := make(chan struct{})
+	go func() {
+		io.Copy(ioutil.Discard, in)
+		stdinDone <- struct{}{}
+	}()
+
+	go func() { // write stdout, loop write to simulate massive data , until conn is closed.
+		for {
+			_, err := io.Copy(out, strings.NewReader("something"))
+			if err != nil && err.Error() != "EOF" {
+				stdoutDone <- struct{}{}
+				return
+			}
+		}
+	}()
+	select {
+	case <-stdinDone:
+		return nil
+	case <-stdoutDone:
+		return nil
+	}
+}
+
+func TestSPDYExecutorStream(t *testing.T) {
+	tests := []struct {
+		name        string
+		options     StreamOptions
+		expectError string
+		attacher    AttachFunc
+	}{
+		{
+			name: "stdoutBlockTest",
+			options: StreamOptions{
+				Stdin:  &fakeMassiveDataPty{},
+				Stdout: &fakeMassiveDataPty{},
+			},
+			expectError: "",
+			attacher:    fakeMassiveDataAttacher,
+		},
+	}
+
+	for _, test := range tests {
+		server := newTestHTTPServer(test.attacher, &test.options)
+
+		err := attach2Server(server.URL, test.options)
+		if err != nil {
+			if err.Error() != test.expectError {
+				t.Errorf("%s: expected [%v], got [%v]", test.name, test.expectError, err.Error())
+			}
+		} else if test.expectError != "" {
+			t.Errorf("%s: expected [%v], got [%v]", test.name, test.expectError, "")
+		}
```

Comment:
probably simpler with something like:

gotError := ""
if err != nil {
gotError = err.Error()
}
if test.expectError != gotError {
t.Errorf(...)
}

Reply:
sure, it's better.
will change as you suggested.

Reply:
updated